x-pack/plugin/esql/compute/src/main/generated-src/org/elasticsearch/compute/aggregation/RateDoubleAggregator.java

```diff
@@ -122,16 +127,25 @@ void ensureCapacity(int groupId) {
             states = bigArrays.grow(states, groupId + 1);
         }
 
+        void adjustBreaker(long bytes) {
+            breaker.addEstimateBytesAndMaybeBreak(bytes, "<<rate aggregation>>");
+            stateBytes += bytes;
+            assert stateBytes >= 0 : stateBytes;
+        }
+
         void append(int groupId, long timestamp, double value) {
             ensureCapacity(groupId);
             var state = states.get(groupId);
             if (state == null) {
+                adjustBreaker(DoubleRateState.bytesUsed(1));
```

Comment:
I think this accounting of memory is good. I just haven't seen this style of accounting for the states of the other aggregate functions (for example for `HllStates.GroupingState` which can also be memory intensive). Is the idea to do this type of accounting for other aggregate functions too?

Reply:
I'm not a big fan of this kind of accounting because it's easier to make a mistake with. But I really like that you can do it up front.

Most of our of other stuff uses BigArrays directly and doesn't need it. But a few could use this trick too.

Reply:
Yes, I agree. This is quite tedious. If we can tolerate the memory usage of one new state (which should not be an issue), then we can integrate this into an ObjectArray.